File: gsd09-web-crawler/html-downloader/src/main/java/htmldownloader/repository/DownloadQueueRepository.java
package htmldownloader.repository;

import htmldownloader.entity.DownloadQueue;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;
import org.springframework.stereotype.Repository;

import java.time.LocalDateTime;
import java.util.List;

@Repository
public interface DownloadQueueRepository extends JpaRepository<DownloadQueue, String> {

    @Query("SELECT dq FROM DownloadQueue dq WHERE dq.status = :status ORDER BY dq.score DESC, dq.createdAt ASC")
    List<DownloadQueue> findByStatusOrderByScoreDescCreatedAtAsc(@Param("status") String status);

    @Query("SELECT dq FROM DownloadQueue dq WHERE dq.crawlingId = :crawlingId AND dq.status = :status")
    List<DownloadQueue> findByCrawlingIdAndStatus(@Param("crawlingId") String crawlingId,
                                                  @Param("status") String status);

    boolean existsByOriginalUrlAndCrawlingId(String originalUrl, String crawlingId);

    /**
     * bfsPath 중복 체크
     */
    boolean existsByBfsPath(String bfsPath);

    /**
     * 특정 크롤링 세션에서 bfsPath 중복 체크
     */
    boolean existsByCrawlingIdAndBfsPath(String crawlingId, String bfsPath);

    /**
     * 1달 이내에 특정 URL이 처리되었는지 확인
     */
    @Query("SELECT COUNT(dq) > 0 FROM DownloadQueue dq WHERE dq.originalUrl = :originalUrl AND dq.createdAt > :oneMonthAgo")
    boolean existsByOriginalUrlWithinOneMonth(@Param("originalUrl") String originalUrl,
                                            @Param("oneMonthAgo") LocalDateTime oneMonthAgo);

    /**
     * 특정 크롤링 세션의 rootDomain 조회
     */
    @Query("SELECT DISTINCT dq.rootDomain FROM DownloadQueue dq WHERE dq.crawlingId = :crawlingId AND dq.depth = 0")
    String findRootDomainByCrawlingId(@Param("crawlingId") String crawlingId);
}

File: sdip09-web-crawler-ktor/html-downloader/src/main/kotlin/htmldownloader/domain/downloadqueue/DownloadQueueRepository.kt
package htmldownloader.domain.downloadqueue

import htmldownloader.util.DateUtils
import org.jetbrains.exposed.sql.*
import org.jetbrains.exposed.sql.SqlExpressionBuilder.eq
import org.jetbrains.exposed.sql.transactions.transaction
import java.time.LocalDateTime

class DownloadQueueRepository(private val database: Database) {

    fun save(downloadQueue: DownloadQueue): DownloadQueue {
        transaction(database) {
            DownloadQueues.insert { row ->
                row[id] = downloadQueue.id
                row[crawlingId] = downloadQueue.crawlingId
                row[originalUrl] = downloadQueue.originalUrl
                row[rootDomain] = downloadQueue.rootDomain
                row[depth] = downloadQueue.depth
                row[bfsPath] = downloadQueue.bfsPath
                row[status] = downloadQueue.status
                row[score] = downloadQueue.score
                row[createdAt] = downloadQueue.createdAt
                row[updatedAt] = downloadQueue.updatedAt
            }
        }
        return downloadQueue
    }

    fun findById(id: String): DownloadQueue? {
        return transaction(database) {
            DownloadQueues.selectAll().where { DownloadQueues.id eq id }
                .map { it.toDownloadQueue() }
                .singleOrNull()
        }
    }

    fun updateStatus(id: String, status: String) {
        transaction(database) {
            DownloadQueues.update({ DownloadQueues.id eq id }) { row ->
                row[DownloadQueues.status] = status
                row[updatedAt] = DateUtils.now()
            }
        }
    }

    fun updateRootDomain(id: String, rootDomain: String) {
        transaction(database) {
            DownloadQueues.update({ DownloadQueues.id eq id }) { row ->
                row[DownloadQueues.rootDomain] = rootDomain
                row[updatedAt] = DateUtils.now()
            }
        }
    }

    fun existsByBfsPath(bfsPath: String): Boolean {
        return transaction(database) {
            DownloadQueues.selectAll().where { DownloadQueues.bfsPath eq bfsPath }
                .count() > 0
        }
    }

    fun existsByCrawlingIdAndBfsPath(crawlingId: String, bfsPath: String): Boolean {
        return transaction(database) {
            DownloadQueues.selectAll().where {
                (DownloadQueues.crawlingId eq crawlingId) and (DownloadQueues.bfsPath eq bfsPath)
            }.count() > 0
        }
    }

    fun existsByOriginalUrlWithinOneMonth(originalUrl: String, oneMonthAgo: LocalDateTime): Boolean {
        return transaction(database) {
            DownloadQueues.selectAll().where {
                (DownloadQueues.originalUrl eq originalUrl) and (DownloadQueues.createdAt greater oneMonthAgo)
            }.count() > 0
        }
    }

    fun findRootDomainByCrawlingId(crawlingId: String): String? {
        return transaction(database) {
            DownloadQueues.selectAll().where {
                (DownloadQueues.crawlingId eq crawlingId) and (DownloadQueues.depth eq 0)
            }.map { it[DownloadQueues.rootDomain] }
                .firstOrNull()
        }
    }

    private fun ResultRow.toDownloadQueue(): DownloadQueue = DownloadQueue(
        id = this[DownloadQueues.id],
        crawlingId = this[DownloadQueues.crawlingId],
        originalUrl = this[DownloadQueues.originalUrl],
        rootDomain = this[DownloadQueues.rootDomain],
        depth = this[DownloadQueues.depth],
        bfsPath = this[DownloadQueues.bfsPath],
        status = this[DownloadQueues.status],
        score = this[DownloadQueues.score],
        createdAt = this[DownloadQueues.createdAt],
        updatedAt = this[DownloadQueues.updatedAt]
    )
}
